tools/cdi-func-test-bad-webserver/main.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"regexp"
+)
+
+func failHEAD(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "HEAD" {
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+	redirect(w, r)
+}
+
+func flaky(w http.ResponseWriter, r *http.Request) {
+	if getCounter()%4 == 3 {
+		// succeed after 10 attempts
+		redirect(w, r)
+		return
+	}
+	w.WriteHeader(http.StatusServiceUnavailable)
+}
+
+func redirect(w http.ResponseWriter, r *http.Request) {
+	re := regexp.MustCompile(`[^/]*$`)
+	requestedFile := re.Find([]byte(r.URL.String()))
+
+	cdiNamespace := os.Getenv("CDI_NAMESPACE")
+	redirectURL := fmt.Sprintf("http://cdi-file-host.%s/%s", cdiNamespace, requestedFile)
+	http.Redirect(w, r, redirectURL, 301)
+}
+
+func initCounter() {
+	newCounterBuf := make([]byte, binary.MaxVarintLen64)
+	_ = binary.PutUvarint(newCounterBuf, 0)
+
+	ioutil.WriteFile("state", newCounterBuf, 0644)
+}
+
+func getCounter() uint64 {
+	counterBuf, err := ioutil.ReadFile("state")
```

Comment:
So I think there is a potential race condition here:
The http server can get 2 requests at the same time, and the counter can get messed up potentially. However I don't think it is an issue in this particular case, just wanted to make you aware of it. A lock and a flush should ensure the values are always incremented correctly.